# Práctica Clase 1 

1) Crear una función capaz de convertir números enteros de base 10 a base 2. Debe recibir como parámetro el número a convertir<br>
Consideraciones:<br> 
a. Tratar de resolverlo sin usar la función format(nro,"b")<br>
b. El pdf "conversion-de-decimal-a-binario.pdf" puede servir de ayuda.


In [11]:
def NumeroBinario(numero):
    '''
    Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
    representación binaria. Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    if type(numero) != int:
        return None # 19; Como se usa esto; 18 tercios; trece

    if numero < 0:
        return None
        
    elif numero == 0:
        return 0

    lista = [] #Para almacenar los restos
    div_resto = 0   
    
    while (numero > 0):
        div_resto = numero % 2 #Resto de la division
        numero = numero // 2 #Division sucesiva por 2
        lista.append(div_resto) #Guardar resto en la lista
    
    numero_binario = '' #Esto vamos a retornar
    for elemento in lista:
        numero_binario = str(elemento) + numero_binario
    
    return int(numero_binario)

In [2]:
def NumeroBinario2(numero):
    '''
    Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
    representación binaria. Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    lista = []
    if (type(numero) == int and numero > -1 ):
      while numero > 0 :
        if numero % 2 == 0:
          lista.insert (0, 0)
        else:
          lista.insert(0, 1)  
        numero = numero // 2
    else:
      return None

    return (lista)


In [3]:
import math

def decimal_a_binario(nro):
    if nro == 0:
        return '0'
    
    binario = ''
    negativo = False
    
    if nro < 0:
        negativo = True
        nro = abs(nro)
    
    while nro > 0:
        residuo = nro % 2
        binario = str(residuo) + binario
        nro = nro // 2
    
    if negativo:
        binario = '-' + binario
    
    return binario 


In [19]:
def NumeroBinario3(numero):
    '''
    Esta función recibe como parámetro un número entero mayor ó igual a cero y lo devuelve en su 
    representación binaria. Debe recibir y devolver un valor de tipo entero.
    En caso de que el parámetro no sea de tipo entero y mayor a -1 retorna nulo.
    '''
    if type(numero) != int:
        return None
    if numero < 0:
        return None
    elif numero == 0:
        return 0

    binario = '' #Tipo str
    div_resto = 0

    while numero > 0:
        div_resto = numero % 2 # Tipo int
        numero = numero // 2  # numero += 2 -> numero = numero + 2 ---> numero //= 2
        binario += str(div_resto)

    binario = binario[::-1]
    return int(binario)

In [20]:
help(help)

Help on _Helper in module _sitebuiltins object:

class _Helper(builtins.object)
 |  Define the builtin 'help'.
 |  
 |  This is a wrapper around pydoc.help that provides a helpful message
 |  when 'help' is typed at the Python interactive prompt.
 |  
 |  Calling help() at the Python prompt starts an interactive help session.
 |  Calling help(thing) prints help for the python object 'thing'.
 |  
 |  Methods defined here:
 |  
 |  __call__(self, *args, **kwds)
 |      Call self as a function.
 |  
 |  __repr__(self)
 |      Return repr(self).
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors defined here:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)



In [4]:
#Ejemplo:
NumeroBinario(29)

11101

In [5]:
#Ejemplo:
NumeroBinario2(29)

[1, 1, 1, 0, 1]

In [6]:
#Ejemplo:
decimal_a_binario(29)

'11101'

In [14]:
NumeroBinario3(29)

11101

2) Convertir de decimal a binario las fracciones 1/2, 1/3, 1/4, 1/5, 1/6, 1/7, 1/8, 1/9. Luego analizar los resultados y observar qué particularidad se encuentra en los mismos. Se puede usar Python o una calculadora, lo importante es ver si hay algo que podemos notar...

In [32]:
def NumeroBinarioFraccionario():
   # if type(decimal) != int:
    #    return None
    decimal = input('Ingrese el valor a trasnformar a binario, debe ser un numero decimal') #'1/3'
    if ((decimal >= 1) or (decimal < 0)):
        return None
    elif decimal == 0:
        return 0
    
    lim_perdiodico = 24
    lista = []

    while decimal > 0 and lim_perdiodico > 0:
        lista.append(int(decimal * 2))
        decimal = decimal * 2 - (int(decimal * 2))
        lim_perdiodico -= 1
    
    numero_binario = '0.'
    for elemento in lista:
        numero_binario = numero_binario + str(elemento)

    return(numero_binario)

In [33]:
NumeroBinarioFraccionario()

ValueError: could not convert string to float: '1/3'

In [27]:
for i in range(2,10):
    print(f'Fraccion 1 / {i} : {1/i} y en binario: {NumeroBinarioFraccionario(1/i)}')

Fraccion 1 / 2 : 0.5 y en binario: 0.1
Fraccion 1 / 3 : 0.3333333333333333 y en binario: 0.010101010101010101010101
Fraccion 1 / 4 : 0.25 y en binario: 0.01
Fraccion 1 / 5 : 0.2 y en binario: 0.001100110011001100110011
Fraccion 1 / 6 : 0.16666666666666666 y en binario: 0.001010101010101010101010
Fraccion 1 / 7 : 0.14285714285714285 y en binario: 0.001001001001001001001001
Fraccion 1 / 8 : 0.125 y en binario: 0.001
Fraccion 1 / 9 : 0.1111111111111111 y en binario: 0.000111000111000111000111


Salida esperada:
* Fraccion 1 / 2 :  0.5  y En binario:  0.1
* Fraccion 1 / 3 :  0.3333333333333333  y En binario:  0.010101010101010101010101
* Fraccion 1 / 4 :  0.25  y En binario:  0.01
* Fraccion 1 / 5 :  0.2  y En binario:  0.001100110011001100110011
* Fraccion 1 / 6 :  0.16666666666666666  y En binario:  0.001010101010101010101010
* Fraccion 1 / 7 :  0.14285714285714285  y En binario:  0.001001001001001001001001
* Fraccion 1 / 8 :  0.125  y En binario:  0.001
* Fraccion 1 / 9 :  0.1111111111111111  y En binario:  0.000111000111000111000111

In [26]:
for i in range(2,10):
    print(f'Fraccion 1 / {i} : {1/i} y en binario: {NumeroBinarioFraccionario(1/i)}')

Fraccion 1 / 2 : 0.5 y en binario: 0.1
Fraccion 1 / 3 : 0.3333333333333333 y en binario: 0.010101010101010101010101
Fraccion 1 / 4 : 0.25 y en binario: 0.01
Fraccion 1 / 5 : 0.2 y en binario: 0.001100110011001100110011
Fraccion 1 / 6 : 0.16666666666666666 y en binario: 0.001010101010101010101010
Fraccion 1 / 7 : 0.14285714285714285 y en binario: 0.001001001001001001001001
Fraccion 1 / 8 : 0.125 y en binario: 0.001
Fraccion 1 / 9 : 0.1111111111111111 y en binario: 0.000111000111000111000111


In [35]:
for e in range(2,10):
    print('Fraccion 1 / {0} : {1} y en binario: {2}'.format(e, 1/e, NumeroBinarioFraccionario(1/e)))


TypeError: NumeroBinarioFraccionario() takes 0 positional arguments but 1 was given

In [34]:
def NumeroBinarioFraccion(numero):
    if type(numero)!= float:
        return "El valor ingresado debe ser entero"
    elif numero<0:
        return "El valor ingresado debe ser positivo"
    else:
        binario = '0.'
        auxiliar = numero
        while auxiliar < 2:
            auxiliar= auxiliar*2
            if auxiliar < 1:
                binario = binario + str(0)
            elif auxiliar > 1:
                binario =binario + str(1)
                auxiliar=auxiliar - 1
            else:
                binario =binario + str(1)
                break
        return  binario


In [37]:
for e in range(2,10):
    print('Fraccion 1 / {0} : {1} y en binario: {2}'.format(e, 1/e, NumeroBinarioFraccion(1/e)))


Fraccion 1 / 2 : 0.5 y en binario: 0.1
Fraccion 1 / 3 : 0.3333333333333333 y en binario: 0.010101010101010101010101010101010101010101010101010101
Fraccion 1 / 4 : 0.25 y en binario: 0.01
Fraccion 1 / 5 : 0.2 y en binario: 0.001100110011001100110011001100110011001100110011001101
Fraccion 1 / 6 : 0.16666666666666666 y en binario: 0.0010101010101010101010101010101010101010101010101010101
Fraccion 1 / 7 : 0.14285714285714285 y en binario: 0.001001001001001001001001001001001001001001001001001001
Fraccion 1 / 8 : 0.125 y en binario: 0.001
Fraccion 1 / 9 : 0.1111111111111111 y en binario: 0.000111000111000111000111000111000111000111000111000111


In [39]:
def FraccionBinaria(numerador, denominador):
    
    if type(numerador) != int or type(denominador) != int:
        return None
    
    elif numerador <= 0 or denominador <= 0:
        return None
    
    fraccion = numerador / denominador
    entero = int(fraccion)
    decimal = round(fraccion % 1,20)
    #print(entero, decimal)

    binario_entero = []
    division_entero = True

    while len(binario_decimal) < 21:
        binario_decimal.append(int(decimal*2))
        if decimal*2 == 1.0:
            break
        elif decimal*2 > 1:
            decimal = (decimal*2)-1
        else:
            decimal = decimal*2
    binario_decimal = ''.join(map(str,binario_decimal))

    resultado = float(binario_entero + '.' + binario_decimal)
        
    #print(binario_entero)
    #print(binario_decimal)
    return resultado



In [41]:
for e in range(2,10):
    print('Fraccion 1 / {0} : {1} y en binario: {2}'.format(e, 1/e, FraccionBinaria(1,e)))


UnboundLocalError: cannot access local variable 'binario_decimal' where it is not associated with a value

In [42]:
def FraccionBinario(fraccion, k):
    '''Esta función transforma un número decimal en número binario.
    param: 
    fraccion: Número decimal a transformar
    k: número de dígitos o precisión del decimal''' 
    
    entero = int(fraccion)
    decimal = fraccion - entero
    
    #Transformo primero la parte entera de la fracción
    lista_resto = [] 
    while entero > 0: 
        lista_resto.append(entero % 2) 
        entero = entero//2
    binario_entero = ''.join(map(str,lista_resto[::-1]))

    #Ahora transformo la parte decimal de la fracción
    lista_enteros = []
    while k > 0:  
        decimal = decimal * 2
        ent = int(decimal)
        lista_enteros.append(ent)
        decimal = decimal - ent
        k -=1
    binario_decimal = ''.join(map(str,lista_enteros))

    return binario_entero+'.'+binario_decimal


In [44]:
for e in range(2,10):
    print('Fraccion 1 / {0} : {1} y en binario: {2}'.format(e, 1/e, FraccionBinario(1/e, 24)))


Fraccion 1 / 2 : 0.5 y en binario: .100000000000000000000000
Fraccion 1 / 3 : 0.3333333333333333 y en binario: .010101010101010101010101
Fraccion 1 / 4 : 0.25 y en binario: .010000000000000000000000
Fraccion 1 / 5 : 0.2 y en binario: .001100110011001100110011
Fraccion 1 / 6 : 0.16666666666666666 y en binario: .001010101010101010101010
Fraccion 1 / 7 : 0.14285714285714285 y en binario: .001001001001001001001001
Fraccion 1 / 8 : 0.125 y en binario: .001000000000000000000000
Fraccion 1 / 9 : 0.1111111111111111 y en binario: .000111000111000111000111


In [45]:
def decimal_a_binario(decimal, precision):
    if decimal >= 1 or decimal <= 0:
        return "Error: el número debe ser mayor a cero y menor a uno"

    bits = []  # Lista para almacenar los bits del número binario

    for _ in range(precision):
        
        decimal = decimal * 2
        bit = int(decimal) # Obtenemos la parte entera de la multiplicación
        bits.append(bit)
        
        decimal = decimal - bit # para obtener la parte fraccionaria
    bits.reverse()
    return bits

numero_decimal = 0.625
numero_binario = decimal_a_binario(numero_decimal, 4)  # Especificamos una precisión de 4 bits

print(numero_binario)  # Salida: [0, 1, 0, 1]


[0, 1, 0, 1]


In [47]:
for e in range(2,10):
    print('Fraccion 1 / {0} : {1} y en binario: {2}'.format(e, 1/e, FraccionBinario(1/e, 24)))


Fraccion 1 / 2 : 0.5 y en binario: .100000000000000000000000
Fraccion 1 / 3 : 0.3333333333333333 y en binario: .010101010101010101010101
Fraccion 1 / 4 : 0.25 y en binario: .010000000000000000000000
Fraccion 1 / 5 : 0.2 y en binario: .001100110011001100110011
Fraccion 1 / 6 : 0.16666666666666666 y en binario: .001010101010101010101010
Fraccion 1 / 7 : 0.14285714285714285 y en binario: .001001001001001001001001
Fraccion 1 / 8 : 0.125 y en binario: .001000000000000000000000
Fraccion 1 / 9 : 0.1111111111111111 y en binario: .000111000111000111000111


In [ ]:
nombre = 'GuAAanzo' #Gonzalo
apellido = '' #Valor faltante


In [ ]:
excel = 'Alumnos FT14.xlsx'

open(excel, 'r')

¿Cuantos alumnso hay?

Edad promedio

Cantidad de mujeres y hombres

